docs/0003-typescript-guidance.md

```diff
@@ -0,0 +1,50 @@
+# Introducing TypeScript
+
+## Status
+Accepted
+
+## Context
+TypeScript is a strongly-typed superset of JavaScript that adds optional static type checking, class-based object-oriented programming, and other features to JavaScript.  
+Here are some of the advantages of TypeScript over JavaScript:
+
+### Type safety
+TypeScript helps catch errors at compile-time instead of runtime by adding type annotations to variables, functions, and classes. This can help prevent errors that might occur when dealing with large codebases.
+### Better tooling support
+TypeScript has better tooling support than JavaScript, with features like code navigation, auto-completion, and refactoring tools in popular code editors like Visual Studio Code.
+### Improved code organization
+TypeScript's class-based object-oriented programming model allows developers to organize code in a more structured and maintainable way.
+### Easy adoption
+TypeScript is a superset of JavaScript, which means that any valid JavaScript code is also valid TypeScript code. This makes it easy for developers to adopt TypeScript gradually, without needing to rewrite their entire codebase.
+### Community support
+TypeScript has a growing community of developers who contribute to its development, create libraries and tools, and provide support to other developers. This makes it easier for developers to learn and adopt TypeScript.
+
+## Decision
+### Target Goals
+Prioritize using TypeScript in the following places:
+* New code files
+* Existing API endpoints (and their payloads)
+* Components or Functions take a lot of parameters, or use parameters that are themselves complex objects
+
+Less important are:
+* React components without complex input objects (especially since they are often adequately served with PropTypes)
```

Comment:
[curious] I'm wondering whether having to decide between adding types via `PropTypes` or TypeScript types being dependent on something relatively subjective (i.e., define "complex") could lead to some potential confusion around when to use which, when, and why. In some situations, we may want/need both. For example, in [this Paragon PR to add types to all components](https://github.com/openedx/paragon/pull/2177), both TypeScript types and `PropTypes` are used together. Granted, Paragon's Gatsby documentation website relies on the `PropTypes` object to generate the props API.
<img width="936" alt="image" src="https://github.com/openedx/frontend-build/assets/2828721/5250dc1a-e931-41f7-a622-9e4ba4e94444">